python/whylogs/experimental/core/metrics/embedding_metric.py

```diff
@@ -0,0 +1,129 @@
+
+
+@serializer(type_id=101)
+def serialize(value: np.ndarray) -> MetricComponentMessage:
+    return MetricComponentMessage(serialized_bytes=_serialize_ndarray(value))
+
+
+@deserializer(type_id=101)
+def deserialize(msg: MetricComponentMessage) -> np.ndarray:
+    return _deserialize_ndarray(msg.serialized_bytes)
+
+
+class DistanceFunction(Enum):
+    euclidean = euclidean_distances
+    cosine = cosine_distances
+
+
+@dataclass(frozen=True)
+class EmbeddingConfig(MetricConfig):
+    references: np.ndarray  # columns are reference vectors
+    labels: Optional[List[str]] = None
+    distance_fn: DistanceFunction = DistanceFunction.cosine
+
+    def __post_init__(self) -> None:
+        if len(self.references.shape) != 2:
+            raise ValueError("Embedding reference matrix must be 2 dimensional")
+
+        if self.labels:
+            if len(self.labels) == self.references.shape[1]:
+                raise ValueError(
+                    f"Number of labels ({len(self.labels)}) must match number of reference vectors ({self.references.shape[1]})"
+                )
+
+
+@dataclass
+class EmbeddingMetric(MultiMetric):
+    references: MatrixComponent
+    labels: List[str]
+    distance_fn: DistanceFunction
+
+    def __post_init__(self):
+        submetrics = {
+            f"{label}_distance": {
+                "distribution": StandardMetric.distribution.zero(),
+                "counts": StandardMetric.counts.zero(),
+                "types": StandardMetric.types.zero(),
+            }
+            for label in self.labels
+        }
+        submetrics.update(
+            {
+                "closest": {
+                    "frequent_items": StandardMetric.frequent_items.zero(),
+                    "counts": StandardMetric.counts.zero(),
+                    "types": StandardMetric.types.zero(),
+                }
+            }
+        )
+        super().__init__(submetrics)
+
+    @property
+    def namespace(self) -> str:
+        return "embedding"
+
+    def _update_submetrics(self, submetric: str, data: PreprocessedColumn) -> None:
+        for key in self.submetrics[submetric].keys():
+            self.submetrics[submetric][key].columnar_update(data)
+
+    def columnar_update(self, data: PreprocessedColumn) -> OperationResult:
+        X = data.list.objs  # TODO: throw if not 2D
+        if not X:
+            return OperationResult.ok(0)
+
+        X_ref_dists = self.distance_fn.value(X, self.references.value)
+        X_ref_closest = np.argmin(self.X_ref_dists, axis=1)
+        closest: List[str] = []
+        for i in range(X_ref_dists.shape[1]):
+            closest.append(self.lables[X_ref_closest[i]])
+            self._update_submetrics(f"{i}_distance", PreprocessedColumn.apply(X_ref_dists[i]))
+
+        self._update_submetrics("closest", PreprocessedColumn.apply(closest))
+
+        return OperationResult.ok(1)
+
+    @classmethod
+    def zero(cls, cfg: Optional[EmbeddingConfig] = None) -> "EmbeddingMetric":
```

Comment:
The zero is a little weird without reference vector, can you merge a zero EmbeddingMetric with a properly initialized one? If merging with zero doesn't work, might be least surprising behavior to raise an error.

Reply:
Let's check with @bernease, but I would guess that merging metrics with differently shaped references or different labels should fail. A `zero()`ed metric is fine as long as it was passed a compatible `EmbeddingConfig`. I can hack `merge()` to return the other metric if one was default configured.

Reply:
No thoughts on zero metric, but yes, merging metrics with different references should fail.

Reply:
You might want to allow merging with zero to take the other metric, but sounds good, and raising an error when different references or shape makes sense. Thanks for explaining.